examples/nlp/text_generation_gpt.py

```diff
@@ -0,0 +1,418 @@
+method. This is fixed by our next search utility, top-k search.
+"""
+
+"""
+### Top-K Search
+
+Similar to random search, we sample the next token from the probability distribution
+provided by the model. The only difference is that here, we select out the top `k` most
+probable tokens, and distribute the probabiltiy mass over them before sampling. This way,
+we won't be sampling from low probability tokens, and hence we would have less
+nonsensical words!
+"""
+
+output_tokens = keras_nlp.utils.top_k_search(
+    token_logits_fn,
+    tf.convert_to_tensor(start_tokens),
+    max_length=NUM_TOKENS_TO_GENERATE,
+    k=10,
+    from_logits=True,
+)
+txt = tokenizer.detokenize(output_tokens)
+print(f"Top-K search generated text: \n{txt}\n")
+
+"""
+### Top-P Search
+
+Even with the top-k search, there is something to improve upon. With top-k search, the
+number `k` is fixed, which means it selects the same number of tokens for any probability
+distribution. Consider two scenarios, one where the probability mass is concentrated over
+2 words and another where the probability mass is evenly concentrated across 10. Should
+we choose `k=2` or `k=10`? There is not a one size fits all `k` here.
+
+This is where top-p search comes in! Instead of choosing a `k`, we choose a probability
+`p` that we want the probabilities of the top tokens to sum up to. This way, we can
+dynamically adjust the `k` based on the probability distribution. By setting `p=0.9`, if
+90% of the probability mass is concentrated on the top 2 tokens, we can filter out the
+top 2 tokens to sample from. If instead the 90% is distributed over 10 tokens, it will
+similarly filter out the top 10 tokens to sample from.
+"""
+
+output_tokens = keras_nlp.utils.top_p_search(
+    token_logits_fn,
+    tf.convert_to_tensor(start_tokens),
+    max_length=NUM_TOKENS_TO_GENERATE,
+    p=0.5,
+    from_logits=True,
+)
+txt = tokenizer.detokenize(output_tokens)
+print(f"Top-P search generated text: \n{txt}\n")
+
+"""
+### Using Callbacks
+
+We can also wrap the utilities in a callback, which allows you to print out a prediction
+sequence for every epoch of the model! Here is an example of a callback for top-k search:
+"""
+
+
+class TopKTextGenerator(keras.callbacks.Callback):
+    """A callback to generate text from a trained model using top-k."""
+
+    def __init__(self, k):
+        self.k = k
+
+    def on_epoch_end(self, epoch, logs=None):
+        output_tokens = keras_nlp.utils.top_k_search(
+            token_logits_fn,
+            tf.convert_to_tensor(start_tokens),
+            max_length=NUM_TOKENS_TO_GENERATE,
+            k=self.k,
+            from_logits=True,
+        )
+        txt = tokenizer.detokenize(output_tokens)
+        print(f"Top-K search generated text: \n{txt}\n")
+
+
+text_generation_callback = TopKTextGenerator(k=10)
+# Dummy training loop to demonstrate callback.
+model.fit(ds.take(1), verbose=2, epochs=2, callbacks=[text_generation_callback])
+
+"""
+## Conclusion
```

Comment:
i'm not sure this conclusion section adds too much. we should either replace it with a further reading/where to next section, or remove it entirely